File: src/lib/token-service.ts
import * as SecureStore from 'expo-secure-store';
import { Platform } from 'react-native';

const KEYS = {
  accessToken: 'access_token',
  refreshToken: 'refresh_token',
  tokenExpires: 'token_expires',
  userEmail: 'user_email',
} as const;

// Web fallback: expo-secure-store not available on web
const storage = {
  get: async (key: string): Promise<string | null> => {
    if (Platform.OS === 'web') {
      return localStorage.getItem(key);
    }
    return SecureStore.getItemAsync(key);
  },
  set: async (key: string, value: string): Promise<void> => {
    if (Platform.OS === 'web') {
      localStorage.setItem(key, value);
      return;
    }
    await SecureStore.setItemAsync(key, value);
  },
  delete: async (key: string): Promise<void> => {
    if (Platform.OS === 'web') {
      localStorage.removeItem(key);
      return;
    }
    await SecureStore.deleteItemAsync(key);
  },
};

export const tokenService = {
  getAccessToken: () => storage.get(KEYS.accessToken),
  getRefreshToken: () => storage.get(KEYS.refreshToken),
  getUserEmail: () => storage.get(KEYS.userEmail),

  getTokenExpires: async (): Promise<number> => {
    const val = await storage.get(KEYS.tokenExpires);
    return val ? parseInt(val, 10) : 0;
  },

  saveTokens: async (
    accessToken: string,
    refreshToken: string,
    expiresIn: number,
    email: string,
  ): Promise<void> => {
    const expiresAt = Date.now() + expiresIn * 1000;
    await Promise.all([
      storage.set(KEYS.accessToken, accessToken),
      storage.set(KEYS.refreshToken, refreshToken),
      storage.set(KEYS.tokenExpires, expiresAt.toString()),
      storage.set(KEYS.userEmail, email),
    ]);
  },

  updateAccessToken: async (
    accessToken: string,
    expiresIn: number,
  ): Promise<void> => {
    const expiresAt = Date.now() + expiresIn * 1000;
    await Promise.all([
      storage.set(KEYS.accessToken, accessToken),
      storage.set(KEYS.tokenExpires, expiresAt.toString()),
    ]);
  },

  clearAll: async (): Promise<void> => {
    await Promise.all(
      Object.values(KEYS).map((key) => storage.delete(key)),
    );
  },

  isTokenExpired: async (): Promise<boolean> => {
    const expires = await tokenService.getTokenExpires();
    // 5 min buffer (same as web api-adapter.js)
    return Date.now() > expires - 5 * 60 * 1000;
  },
};

File: lib/services/token_service.dart
import 'package:shared_preferences/shared_preferences.dart';
import 'package:plantapp/constants/api.dart';

/// Token storage using SharedPreferences (UserDefaults on iOS).
/// Survives app reinstalls — unlike Keychain which gets wiped.
class TokenService {
  TokenService._();
  static final instance = TokenService._();

  static const _accessTokenKey = 'access_token';
  static const _refreshTokenKey = 'refresh_token';
  static const _expiresKey = 'token_expires';
  static const _emailKey = 'user_email';

  Future<String?> getAccessToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_accessTokenKey);
  }

  Future<String?> getRefreshToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_refreshTokenKey);
  }

  Future<String?> getUserEmail() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_emailKey);
  }

  Future<int> getTokenExpires() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getInt(_expiresKey) ?? 0;
  }

  Future<bool> isTokenExpired() async {
    final expires = await getTokenExpires();
    return DateTime.now().millisecondsSinceEpoch >= expires - tokenRefreshBufferMs;
  }

  Future<void> saveTokens({
    required String accessToken,
    required String refreshToken,
    required int expiresIn,
    required String email,
  }) async {
    final prefs = await SharedPreferences.getInstance();
    final expiresAt = DateTime.now().millisecondsSinceEpoch + (expiresIn * 1000);
    await Future.wait([
      prefs.setString(_accessTokenKey, accessToken),
      prefs.setString(_refreshTokenKey, refreshToken),
      prefs.setInt(_expiresKey, expiresAt),
      prefs.setString(_emailKey, email),
    ]);
  }

  Future<void> updateAccessToken(String accessToken, int expiresIn) async {
    final prefs = await SharedPreferences.getInstance();
    final expiresAt = DateTime.now().millisecondsSinceEpoch + (expiresIn * 1000);
    await Future.wait([
      prefs.setString(_accessTokenKey, accessToken),
      prefs.setInt(_expiresKey, expiresAt),
    ]);
  }

  Future<void> clearAll() async {
    final prefs = await SharedPreferences.getInstance();
    await Future.wait([
      prefs.remove(_accessTokenKey),
      prefs.remove(_refreshTokenKey),
      prefs.remove(_expiresKey),
      prefs.remove(_emailKey),
    ]);
  }
}
